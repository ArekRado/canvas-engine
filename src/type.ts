import { Vector2D } from '@arekrado/vector-2d'
import { GlobalSystem, System } from './system/createSystem'
import { TimingFunction } from './util/bezierFunction'

export type Dictionary<Value> = { [key: string]: Value }

export type Guid = string

export type Component<Data> = {
  entity: Entity
  name: string
} & Data

export type CollideType = {
  type: 'box' | 'circle'
  entity: Entity
}

export type CollideBox = Component<{
  size: Vector2D
  position: Vector2D
  collisions: CollideType[]
}>

export type CollideCircle = Component<{
  radius: number
  position: Vector2D
  collisions: CollideType[]
}>

export type Transform = Component<{
  rotation: number
  fromParentRotation: number
  scale: Vector2D
  fromParentScale: Vector2D
  position: Vector2D
  fromParentPosition: Vector2D
  parent?: Entity
}>

export type Blueprint = Component<{
  id: Guid
}>

export type AnimationProperty = {
  path: string
  component: keyof State['component']
  entity: Entity
  index?: number
}

export type AnimationValueRangeNumber = {
  type: 'Number'
  value: Vector2D
}

export type AnimationValueRangeVector2D = {
  type: 'Vector2D'
  value: [Vector2D, Vector2D]
}

export type AnimationValueRange =
  | AnimationValueRangeNumber
  | AnimationValueRangeVector2D

export type WrapMode =
  //When time reaches the end of the animation clip, the clip will automatically stop playing and time will be reset to beginning of the clip.
  | 'Once'
  // When time reaches the end of the animation clip, time will continue at the beginning.
  | 'Loop'
  // When time reaches the end of the animation clip, time will ping pong back between beginning and end.
  | 'PingPong'
  // Plays back the animation. When it reaches the end, it will keep playing the last frame and never stop playing.
  | 'ClampForever'

export type Keyframe = {
  duration: number
  timingFunction: TimingFunction
  valueRange: AnimationValueRange
}

export type Animation = Component<{
  keyframes: Keyframe[]
  isPlaying: boolean
  isFinished: boolean
  currentTime: number
  property: AnimationProperty
  wrapMode: WrapMode
}>

export type SpriteSrc = string

export type Sprite = Component<{ src: SpriteSrc }>

export type AnimatedProperty = {
  path: string
  type: 'number' | 'Vector2D'
}

export type Entity = {
  id: Guid
  name: string
  persistOnSceneChange: boolean
}

export type Time = {
  previousTimeNow: number
  timeNow: number
  delta: number
}

export type AssetSprite = {
  src: string
  name: string
}

export type AssetBlueprint = {
  name: string
  entity: Entity
  data: Dictionary<Component<any>>
}

export type Asset = {
  sprite: AssetSprite[]
  blueprint: AssetBlueprint[]
}

export type Mouse = {
  buttons: number
  position: Vector2D
}

/* blueprint: Belt.Map.String.t({
      connectedEntites: []
    }), */
// TODO
// event
// scene
export type State = {
  entity: Entity[]
  component: Dictionary<Dictionary<Component<any>>> & {
    sprite: Dictionary<Sprite>
    transform: Dictionary<Transform>
    animation: Dictionary<Animation>
    collideBox: Dictionary<CollideBox>
    collideCircle: Dictionary<CollideCircle>
  }
  system: Dictionary<System<any> | GlobalSystem>
  asset: Asset
  mouse: Mouse
  time: Time
  isDebugInitialized: boolean
  isDrawEnabled: boolean
}
