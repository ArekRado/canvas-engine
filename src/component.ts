import { Guid, State } from './type'
import { Component, Dictionary } from './type'

export enum componentName {
  sprite = 'sprite',
  animation = 'animation',
  collideBox = 'collideBox',
  collideCircle = 'collideCircle',
  blueprint = 'blueprint',
  mouseInteraction = 'mouseInteraction',
}

type SetComponentParams<Data> = {
  state: State
  data: Component<Data>
}
export const setComponent = <Data>(
  name: string,
  { state, data }: SetComponentParams<Data>,
): State => {
  const newState = {
    ...state,
    component: {
      ...state.component,
      [name]: {
        ...state.component[name],
        [data.entityId]: data,
      },
    },
  }

  if (
    state.system[name] !== undefined &&
    (state.component[name] === undefined ||
      state.component[name][data.entityId] === undefined)
  ) {
    return state.system[name].create({ state: newState, component: data })
  }

  return newState
}

type RemoveComponent = (
  name: string,
  params: {
    entityId: Guid
    state: State
  },
) => State
export const removeComponent: RemoveComponent = (name, { entityId, state }) => {
  const { [entityId]: _, ...dictionaryWithoutComponent } = state.component[
    name
  ] as Dictionary<Component<any>>

  const newState = {
    ...state,
    component: {
      ...state.component,
      [name]: dictionaryWithoutComponent,
    },
  }

  const component = getComponent(name, { state, entityId })

  if (newState.system[name]) {
    return newState.system[name].remove({ state: newState, component })
  }

  return newState
}

export const getComponent = <Data>(
  name: string,
  {
    entityId,
    state,
  }: {
    entityId: Guid
    state: State
  },
): Component<Data> | undefined => {
  const c: Dictionary<Component<Data>> = state.component[name]
  return c ? (c[entityId] as Component<Data> | undefined) : undefined
}
