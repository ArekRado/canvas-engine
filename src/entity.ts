import { v4 } from 'uuid'
import { Entity, Guid, AnyState } from './type'

export const createEntity = ({ name }: { name: string }): Entity =>
  name ? `${name}---${v4()}` : v4()

type SetEntity = <State extends AnyState>(params: {
  entity: Entity
  state: State
}) => State

export const setEntity: SetEntity = ({ entity, state }) => ({
  ...state,
  entity: {
    ...state.entity,
    [entity]: entity,
  },
})

type GetEntity = <State extends AnyState>(params: {
  entity: Guid
  state: State
}) => Entity | undefined
export const getEntity: GetEntity = ({ entity, state }) => state.entity[entity]

type RemoveEntity = <State extends AnyState>(params: {
  entity: Guid
  state: State
  removeComponent: ({
    name,
    entity,
    state,
  }: {
    name: string
    entity: string
    state: State
  }) => State
}) => State
export const removeEntity: RemoveEntity = ({
  entity,
  state,
  removeComponent,
}) => {
  const { [entity]: _, ...stateWithoutEntity } = state.entity

  const newState = {
    ...state,
    entity: stateWithoutEntity,
  }

  const v1 = Object.keys(newState.component).reduce(
    (state, name) =>
      removeComponent({
        name,
        state,
        entity,
      }),
    newState,
  )

  return v1
}
