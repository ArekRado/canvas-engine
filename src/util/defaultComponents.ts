import { vectorZero } from '@arekrado/vector-2d'
import {
  Camera,
  GetDefaultComponent,
  Mouse,
  MouseInteraction,
  RigidBody,
} from '../type'
import { Animation, Keyboard, Transform, Collider } from '../type'

export const defaultAnimation: GetDefaultComponent<
  Animation.AnimationComponent
> = (data = {}) => ({
  isPlaying: false,
  isFinished: false,
  deleteWhenFinished: false,
  currentTime: 0,
  properties: [],
  wrapMode: Animation.WrapMode.once,
  timingMode: Animation.TimingMode.smooth,
  ...data,
})

export const defaultCollider: GetDefaultComponent<Collider> = (data = {}) => ({
  position: vectorZero(),
  _collisions: [],
  data: [],
  layers: [],
  ...data,
})
export const defaultRigidBody: GetDefaultComponent<RigidBody> = (
  data = {},
) => ({
  mass: 0,
  friction: 0,
  velocity: vectorZero(),
  force: vectorZero(),
  ...data,
})

export const defaultMouseInteraction: GetDefaultComponent<MouseInteraction> = (
  data,
) => ({
  clickSpeed: 200,
  doubleClickSpeed: 200,
  isClicked: false,
  isDoubleClicked: false,
  isMouseOver: false,
  isMouseEnter: false,
  isMouseLeave: false,
  ...data,
})

export const defaultCamera: GetDefaultComponent<Camera> = (data = {}) => ({
  position: vectorZero(),
  distance: 1,
  bottom: 0,
  top: 0,
  left: 0,
  right: 0,
  ...data,
})

export const defaultTransform: GetDefaultComponent<Transform> = (
  data = {},
) => ({
  rotation: vectorZero(),
  fromParentRotation: vectorZero(),
  scale: [1, 1],
  fromParentScale: [1, 1],
  position: vectorZero(),
  fromParentPosition: vectorZero(),
  ...data,
})

export const defaultMouse: GetDefaultComponent<Mouse> = (data = {}) => ({
  buttons: 0,
  position: [0, 0],
  isButtonUp: false,
  isButtonDown: false,
  isMoving: false,
  lastClick: {
    timestamp: -1,
    buttons: 0,
  },
  wheel: {
    deltaMode: 0,
    deltaX: 0,
    deltaY: 0,
    deltaZ: 0,
  },
  ...data,
})

export const defaultKeyboard: GetDefaultComponent<Keyboard> = (data = {}) => ({
  keys: {},
  ...data,
})
