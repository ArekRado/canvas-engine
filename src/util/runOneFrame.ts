import { timeEntity } from '../system/time/time'
import { getTime, updateTime } from '../system/time/timeCrud'
import { AnyState } from '../type'

export const FIXED_TICK_TIME = 1 // 1ms

let _moduloTimeBuffer = 0 // in situation when delta is 15.65 fixedTick will be triggered 15 times, next frame will use this value to increase amount of calls

export const _resetModuloTimeBuffer = () => (_moduloTimeBuffer = 0)

const getFixedTickAmount = (state: AnyState): number => {
  const delta =
    getTime({
      entity: timeEntity,
      state,
    })?.delta ?? 0

  let fixedTickLoops = Math.floor(delta / FIXED_TICK_TIME)
  const modulo = delta % FIXED_TICK_TIME

  if (modulo + _moduloTimeBuffer >= FIXED_TICK_TIME) {
    fixedTickLoops += 1
    _moduloTimeBuffer -= FIXED_TICK_TIME
  } else {
    _moduloTimeBuffer += modulo
  }

  return fixedTickLoops
}

export const runOneFrame = <State extends AnyState = AnyState>({
  state,
}: {
  state: State
}): State => {
  const fixedTickLoops = getFixedTickAmount(state)

  const allSystems = [...state.system, ...state.globalSystem]
    .concat()
    .sort((a, b) => (a.priority > b.priority ? 1 : -1))

  const timeBeforeFixedTicks = getTime({
    entity: timeEntity,
    state,
  })

  // Loop for fixedTicks
  Array.from({ length: fixedTickLoops }).forEach(() => {
    // Fixed tick has to have delta equal to FIXED_TICK_TIME
    state = updateTime({
      state,
      entity: timeEntity,
      update: (time) => ({
        delta: FIXED_TICK_TIME,
        timeNow: time.timeNow + FIXED_TICK_TIME,
        dataOverwrite: {
          ...time.dataOverwrite,
          delta: FIXED_TICK_TIME,
          timeNow: time.timeNow + FIXED_TICK_TIME,
        },
      }),
    }) as State

    for (let i = 0; i < allSystems.length; i++) {
      const system = allSystems[i]
      if (system.fixedTick) {
        state = system.fixedTick({ state }) as State
      }
    }
  })

  // reset time after fixed updates. It's the easiest way to return to previous time values without worrying about modulo
  state = updateTime({
    state,
    entity: timeEntity,
    update: () => ({
      ...timeBeforeFixedTicks,
    }),
  }) as State

  // Loop for normal ticks
  for (let i = 0; i < allSystems.length; i++) {
    const system = allSystems[i]
    if (system.tick) {
      state = system.tick({ state }) as State
    }
  }

  return state
}
