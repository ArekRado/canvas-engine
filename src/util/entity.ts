import { v4 } from 'uuid'
import { Dictionary, Entity, Guid, State } from '../type'
import { removeComponent } from '../component'
import { vector, Vector2D, vectorZero } from '@arekrado/vector-2d'

type Generate = (
  name: string,
  options?: Partial<{
    persistOnSceneChange: boolean
    rotation: number
    fromParentRotation: number
    scale: Vector2D
    fromParentScale: Vector2D
    position: Vector2D
    fromParentPosition: Vector2D
    parentId?: Guid
  }>,
) => Entity
export const generateEntity: Generate = (
  name: string,
  options = {},
): Entity => ({
  name,
  id: v4(),
  persistOnSceneChange: options.persistOnSceneChange || false,
  rotation: options.rotation || 0,
  fromParentRotation: options.fromParentRotation || 0,
  scale: options.scale || vector(1, 1),
  fromParentScale: options.fromParentScale || vector(1, 1),
  position: options.position || vectorZero(),
  fromParentPosition: options.fromParentPosition || vectorZero(),
  parentId: options.parentId,
})

type SetEntity = (params: { entity: Entity; state: State }) => State

export const setEntity: SetEntity = ({ entity, state }) => ({
  ...state,
  entity: {
    ...state.entity,
    [entity.id]: entity,
  },
})

type GetEntity = (params: {
  entityId: Guid
  state: State
}) => Entity | undefined

export const getEntity: GetEntity = ({ entityId, state }) =>
  state.entity[entityId]

type RemoveEntity = (params: { entityId: Guid; state: State }) => State
export const removeEntity: RemoveEntity = ({ entityId, state }) => {
  const { [entityId]: _, ...stateWithoutEntity } = state.entity

  const newState = {
    ...state,
    entity: stateWithoutEntity,
  }

  const v1 = Object.keys(newState.component).reduce(
    (state, name) =>
      removeComponent(name, {
        state,
        entityId,
      }),
    newState,
  )

  return v1
}
