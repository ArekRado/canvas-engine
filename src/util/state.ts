import { colliderSystem } from '../system/collider/collider'
import { transformSystem } from '../system/transform/transform'
import { componentName } from '../component/componentName'
import { timeSystem } from '../system/time/time'
import { mouseInteractionSystem } from '../system/mouseInteraction/mouseInteraction'
import { mouseSystem } from '../system/mouse/mouse'
import { keyboardSystem } from '../system/keyboard/keyboard'
import { cameraSystem } from '../system/camera/camera'
import { animationSystem } from '../system/animation/animation'
import { meshSystem } from '../system/mesh/mesh'
import { materialSystem } from '../system/material/material'
import { eventSystem } from '../event'
import { rigidBodySystem } from '../system/rigidBody/rigidBody'
import { AnyState, InternalInitialState } from '../type'
import { Scene, WebGLRenderer } from 'Three'

let sceneRef: Scene | undefined
let rendererRef: WebGLRenderer | undefined

// export const setScene = () => {
//   sceneRef = new Scene()
// }
// export const getScene = () => sceneRef

// export const setRenderer = (containerId: string) => {
//   const canvas = document.getElementById(containerId) as HTMLCanvasElement
//   rendererRef = new WebGLRenderer({ canvas })
// }
// export const getRenderer = () => rendererRef

export function scene() {
  // let sceneRef: Scene | undefined

  return {
    get: function () {
      return sceneRef
    },
    set: function () {
      sceneRef = new Scene()
    },
  }
}

export function renderer() {
  // let rendererRef: WebGLRenderer | undefined

  return {
    get: function () {
      return rendererRef
    },
    set: function (containerId: string) {
      const canvas = document.getElementById(containerId) as HTMLCanvasElement
      rendererRef = new WebGLRenderer({ canvas })
    },
  }
}

export const getInitialState = (): InternalInitialState => ({
  entity: {},
  component: {
    [componentName.animation]: {},
    [componentName.collider]: {},
    [componentName.mouseInteraction]: {},
    [componentName.time]: {},
    [componentName.camera]: {},
    [componentName.transform]: {},
    [componentName.mouse]: {},
    [componentName.keyboard]: {},
    [componentName.mesh]: {},
    [componentName.material]: {},
    [componentName.rigidBody]: {},
  },
  globalSystem: [],
  system: [],
  animationFrame: -1,
})

export const getSystems = ({
  state,
  document,
  containerId,
}: {
  state: AnyState
  renderer?: WebGLRenderer
  document?: Document
  containerId?: string
}): InternalInitialState => {
  let internatlState = state as InternalInitialState

  internatlState = eventSystem(internatlState) as InternalInitialState
  internatlState = timeSystem(internatlState) as InternalInitialState
  internatlState = cameraSystem(internatlState) as InternalInitialState
  internatlState = transformSystem(internatlState) as InternalInitialState
  internatlState = colliderSystem(internatlState) as InternalInitialState
  internatlState = animationSystem(internatlState) as InternalInitialState
  internatlState = mouseInteractionSystem(
    internatlState,
  ) as InternalInitialState
  internatlState = materialSystem(internatlState) as InternalInitialState
  internatlState = meshSystem(internatlState) as InternalInitialState
  internatlState = rigidBodySystem(internatlState) as InternalInitialState

  if (containerId) {
    internatlState = mouseSystem({
      state: internatlState,
      document: document ?? window.document,
      containerId,
    }) as InternalInitialState
    internatlState = keyboardSystem({
      state: internatlState,
      document: document ?? window.document,
      containerId,
    }) as InternalInitialState
  }

  return internatlState
}

export const getState = <State extends AnyState = AnyState>({
  state,
  document,
  containerId,
}: {
  state?: State
  document?: Document
  window?: Window
  containerId: string
}): InternalInitialState => {
  if (window && document) {
    renderer().set(containerId)
    scene().set()

    // const renderer = new WebGLRenderer({ canvas })
    // const scene = new Scene()

    const rendererRef = renderer().get()
    const sceneRef = scene().get()

    if (rendererRef && sceneRef) {
      rendererRef.setSize(window.innerWidth, window.innerHeight)
      document.body.appendChild(rendererRef.domElement)
    }
  }

  return getSystems({
    state: state || getInitialState(),
    document,
    containerId,
  })
}
