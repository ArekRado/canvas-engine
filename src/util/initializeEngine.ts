import { initialize as initializePixi } from '../util/pixiDraw'
import { initialize as initializeIO } from '../system/io'

export const initializeEngine = async (params?: { containerId?: string }) => {
  const containerId = params?.containerId || 'canvas-engine'

  const isContainerAlreadyExist = document.getElementById(containerId) !== null
  const body = document.body

  if (body && isContainerAlreadyExist === false) {
    const gameContainer = document.createElement('div')
    gameContainer.setAttribute('id', containerId)
    body.appendChild(gameContainer)
  }

  initializeIO()
  await initializePixi(containerId)
}
