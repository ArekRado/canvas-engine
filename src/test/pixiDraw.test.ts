import 'jest-canvas-mock'
import 'jest-webgl-canvas-mock'
import 'regenerator-runtime/runtime'

import { initialState } from '../util/state'
import { runOneFrame } from '../util/runOneFrame'
import { initializeEngine } from '../util/initializeEngine'
import { generateEntity, setEntity } from '../util/entity'
import { componentName, setComponent } from '../component'
import { collideBox, sprite } from '../util/defaultComponents'
import { vector } from '@arekrado/vector-2d'
import { CollideBox, Sprite } from '../type'

describe('pixiDraw', () => {
  beforeEach(() => {
    document.body.innerHTML = ''

    Element.prototype.getBoundingClientRect = jest.fn(() => ({
      width: 100,
      height: 100,
      top: 0,
      left: 0,
      bottom: 0,
      right: 0,
      x: 0,
      y: 0,
      toJSON: () => {},
    }))
  })

  it('initializeEngine should mount canvas with pixi and create div with default id', async () => {
    await initializeEngine()

    expect(document.body.innerHTML).toMatchSnapshot()
  })

  it('initializeEngine should mount canvas with pixi and create div with id passed as param', async () => {
    await initializeEngine({ containerId: 'containerId' })

    expect(document.body.innerHTML).toMatchSnapshot()
  })

  it('initializeEngine should not create new container when it already exists', async () => {
    document.body.innerHTML = '<div id="containerId"/>'

    await initializeEngine({ containerId: 'containerId' })

    expect(document.body.innerHTML).toMatchSnapshot()
  })

  it('initializeEngine should mount canvas with pixi and create div with proper id', async () => {
    await initializeEngine()

    const entity1 = generateEntity('e1', {
      position: vector(50, 50),
    })
    const entity2 = generateEntity('e2', {
      position: vector(50, 50),
    })

    const v1 = setEntity({
      entity: entity1,
      state: initialState,
    })

    const v2 = setEntity({
      entity: entity1,
      state: v1,
    })

    const v3 = setComponent<CollideBox>(componentName.collideBox, {
      state: v2,
      data: collideBox({
        entityId: entity1.id,
        size: vector(1.5, 1.5),
        position: vector(1, 1),
      }),
    })

    const v4 = setComponent<Sprite>(componentName.sprite, {
      state: v3,
      data: sprite({
        entityId: entity2.id,
        src: './asset/testImage.png',
      }),
    })

    runOneFrame({ state: v4, timeNow: 0 })

    const canvas = document.querySelector('canvas')

    if (canvas) {
      // TODO: why it's empty?
      expect(canvas.toDataURL('image/jpeg', 1.0)).toBe(
        'data:image/jpeg;base64,00',
      )
    }
  })

  it('should not throw error when sprite src has been changed', async () => {
    await initializeEngine()

    const entity = generateEntity('')

    let state = setEntity({
      entity,
      state: initialState,
    })

    state = setEntity({
      entity,
      state,
    })

    state = setComponent<Sprite>(componentName.sprite, {
      state,
      data: sprite({
        entityId: entity.id,
        src: 'data:image/png;base64,01',
      }),
    })

    runOneFrame({ state, timeNow: 0 })

    state = setComponent<Sprite>(componentName.sprite, {
      state,
      data: sprite({
        entityId: entity.id,
        src: 'data:image/png;base64,02',
      }),
    })

    runOneFrame({ state, timeNow: 0 })

    expect(true).toBeTruthy()
  })
})
