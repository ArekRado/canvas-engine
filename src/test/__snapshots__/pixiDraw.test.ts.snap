// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pixiDraw initializeEngine should mount canvas with pixi and create div with default id 1`] = `"<div id=\\"canvas-engine\\"><canvas width=\\"100\\" height=\\"100\\"></canvas></div>"`;

exports[`pixiDraw initializeEngine should mount canvas with pixi and create div with id passed as param 1`] = `"<div id=\\"containerId\\"><canvas width=\\"100\\" height=\\"100\\"></canvas></div>"`;

exports[`pixiDraw initializeEngine should not create new container when it already exists 1`] = `"<div id=\\"containerId\\"><canvas width=\\"100\\" height=\\"100\\"></canvas></div>"`;
